.github/workflows/ci.yml

```diff
@@ -9,60 +9,102 @@ on:
+        uses: docker/setup-buildx-action@v3
+      - name: Cache RUN --mount=type=cache /var/cache/apt
+        uses: actions/cache@v3
         with:
-          python-version-file: '.python-version'
-      - name: pip cache
-        uses: actions/cache@v4
+          path: var-cache-apt
+          key: var-cache-apt-${{ hashFiles('Dockerfile') }}
+          restore-keys: |
+            var-cache-apt-
+      - name: Cache RUN --mount=type=cache /var/lib/apt
+        uses: actions/cache@v3
         with:
-          path: ~/.cache/pip
-          key: pip-${{ runner.os }}
-      - name: Cache built Python environment
-        uses: actions/cache@v4
+          path: var-lib-apt
+          key: var-lib-apt-${{ hashFiles('Dockerfile') }}
+          restore-keys: |
+            var-lib-apt-
+      - name: Cache RUN --mount=type=cache /root/.cache/pip
+        uses: actions/cache@v3
+        with:
+          path: root-cache-pip
+          key: root-cache-pip-${{ hashFiles('requirements/*.txt') }}
+          restore-keys: |
+            root-cache-pip-
+      - name: Cache RUN --mount=type=cache /root/.npm
+        uses: actions/cache@v3
+        with:
+          path: root-npm
+          key: root-npm-${{ hashFiles('package.json', 'package-lock.json') }}
+          restore-keys: |
+            root-npm-
+      - name: inject var-cache-apt into docker
+        uses: reproducible-containers/buildkit-cache-dance@v2.1.4
         with:
-          path: ${{ env.pythonLocation }}
-          key: ${{ runner.os }}-${{ env.pythonLocation }}-${{ hashFiles('requirements.txt', 'requirements/*.txt') }}
-      - name: Install Python dependencies
+          cache-source: var-cache-apt
+          cache-target: /var/cache/apt
+          skip-extraction: ${{ steps.cache-var-cache-apt.outputs.cache-hit }}
+      - name: inject var-lib-apt into docker
+        uses: reproducible-containers/buildkit-cache-dance@v2.1.4
+        with:
+          cache-source: var-lib-apt
+          cache-target: /var/lib/apt
+          skip-extraction: ${{ steps.cache-var-lib-apt.outputs.cache-hit }}
+      - name: inject root-cache-pip into docker
+        uses: reproducible-containers/buildkit-cache-dance@v2.1.4
+        with:
+          cache-source: root-cache-pip
+          cache-target: /root/.cache/pip
+          skip-extraction: ${{ steps.cache-root-cache-pip.outputs.cache-hit }}
+      - name: inject root-npm into docker
+        uses: reproducible-containers/buildkit-cache-dance@v2.1.4
+        with:
+          cache-source: root-npm
+          cache-target: /root/.npm
+          skip-extraction: ${{ steps.cache-root-npm.outputs.cache-hit }}
+      - name: Build Docker Image
         run: |
-          pip install -U setuptools wheel pip
-          pip install -r requirements.txt --no-deps
-          pip install -r requirements/dev.txt
-          pip check
+            echo ${{ secrets.GITHUB_TOKEN }} | docker login ghcr.io -u ${{ github.repository_owner }} --password-stdin
```

Comment:
Using stdin here to avoid writing the password to the log? Wouldn't GHA mask the token anyways?

Reply:
This is just following guidance from GHA docs since we are using our own buildx to access gha cache type

https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-with-a-personal-access-token-classic

Reply:
Good enough for me!